File: speech_to_text/csharp/real_time/stream_file/StreamFile.cs
using System;
using System.IO;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

class Realtime
{
    static async Task Main()
    {
        // Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
        string? apiKey = Environment.GetEnvironmentVariable("SONIOX_API_KEY");
        if (string.IsNullOrEmpty(apiKey))
        {
            throw new InvalidOperationException("SONIOX_API_KEY is not set.");
        }
        var websocketUrl = new Uri("wss://stt-rt.soniox.com/transcribe-websocket");

        using (ClientWebSocket ws = new ClientWebSocket())
        {
            // Connect to WebSocket API
            Console.WriteLine("Opening WebSocket connection...");
            await ws.ConnectAsync(websocketUrl, CancellationToken.None);

            // Send start request with correct field names
            var startMessage = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(new
            {
                api_key = apiKey,
                audio_format = "pcm_s16le",
                sample_rate = 16000,
                num_channels = 1,
                model = "stt-rt-preview",
                language_hints = new []{"en", "es"}
            }));
            await ws.SendAsync(new ArraySegment<byte>(startMessage), WebSocketMessageType.Text, true, CancellationToken.None);

            // Start send audio task
            var sendAudioTask = Task.Run(async () =>
            {
                using (FileStream fs = new FileStream("../../coffee_shop.pcm_s16le", FileMode.Open, FileAccess.Read))
                {
                    Console.WriteLine("Transcription started.");
                    var buffer = new byte[3840];
                    int bytesRead;
                    while ((bytesRead = await fs.ReadAsync(buffer, 0, buffer.Length)) > 0)
                    {
                        if (ws.State != WebSocketState.Open)
                        {
                            return;
                        }

                        var audioChunk = new ArraySegment<byte>(buffer, 0, bytesRead);
                        await ws.SendAsync(audioChunk, WebSocketMessageType.Binary, true, CancellationToken.None);

                        await Task.Delay(120);
                    }

                    // Send end of file (empty message)
                    await ws.SendAsync(new ArraySegment<byte>(Array.Empty<byte>()), WebSocketMessageType.Text, true, CancellationToken.None);
                }
            });

            try
            {
                // Receive text
                var buffer = new byte[8192];

                var finalText = "";

                while (true)
                {
                    if (ws.State != WebSocketState.Open)
                    {
                        break;
                    }

                    var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);

                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                        break;
                    }

                    if (result.MessageType == WebSocketMessageType.Text)
                    {
                        string message = Encoding.UTF8.GetString(buffer, 0, result.Count);
                        var res = JsonSerializer.Deserialize<JsonElement>(message);

                        // Handle any possible transcription error
                        if (res.TryGetProperty("error_code", out var errorCode))
                        {
                            res.TryGetProperty("error_message", out var errorMessage);
                            Console.WriteLine("");
                            Console.WriteLine("Error {0}: {1}", errorCode.GetInt32(), errorMessage.GetString());
                            await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                            break;
                        }

                        var nonFinalText = "";

                        // Process tokens
                        if (res.TryGetProperty("tokens", out var tokens))
                        {
                            foreach (var token in tokens.EnumerateArray())
                            {
                                if (token.TryGetProperty("text", out var text))
                                {
                                    if (token.TryGetProperty("is_final", out var isFinal))
                                    {
                                        if (isFinal.GetBoolean()) {
                                            finalText += text.GetString();
                                        } else {
                                            nonFinalText += text.GetString();
                                        }
                                    }
                                }
                            }
                        }

                        Console.Clear();
                        Console.SetCursorPosition(0, 0);
                        Console.Write(finalText);
                        Console.ForegroundColor = ConsoleColor.Blue;
                        Console.Write(nonFinalText);
                        Console.ResetColor();

                        if (res.TryGetProperty("finished", out var finished) && finished.GetBoolean())
                        {
                            // Close the WebSocket
                            Console.WriteLine("\nTranscription done.");
                            await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                            break;
                        }
                    }
                }

                Console.WriteLine();
            }
            finally
            {
                await sendAudioTask;
            }
        }
    }
}

File: speech_to_text/go/real_time/stream_file/stream_file.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"time"

	"github.com/gorilla/websocket"
)

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
var apiKey = os.Getenv("SONIOX_API_KEY")
var websocketUrl = "wss://stt-rt.soniox.com/transcribe-websocket"
var fileToTranscribe = "coffee_shop.pcm_s16le"

type StartRequest struct {
	ApiKey        string   `json:"api_key"`
	Model         string   `json:"model"`
	AudioFormat   string   `json:"audio_format,omitempty"`
	SampleRate    int      `json:"sample_rate,omitempty"`
	NumChannels   int      `json:"num_channels,omitempty"`
	LanguageHints []string `json:"language_hints,omitempty"`
}

type Token struct {
	Text       string  `json:"text"`
	StartMs    int     `json:"start_ms"`
	EndMs      int     `json:"end_ms"`
	Confidence float32 `json:"confidence"`
	IsFinal    bool    `json:"is_final"`
}

type Response struct {
	Tokens       []Token `json:"tokens"`
	Finished     bool    `json:"finished,omitempty"`
	ErrorCode    int     `json:"error_code,omitempty"`
	ErrorMessage string  `json:"error_message,omitempty"`
}

func main() {
	file, err := os.Open(fileToTranscribe)
	if err != nil {
		panic(err)
	}

	fmt.Println("Opening WebSocket connection...")

	conn, _, err := websocket.DefaultDialer.Dial(websocketUrl, nil)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	// Send start request

	startRequest, err := json.Marshal(StartRequest{
		ApiKey:        apiKey,
		AudioFormat:   "pcm_s16le",
		SampleRate:    16000,
		NumChannels:   1,
		Model:         "stt-rt-preview-v2",
		LanguageHints: []string{"en", "es"},
	})
	if err != nil {
		panic(err)
	}
	if err := conn.WriteMessage(websocket.TextMessage, startRequest); err != nil {
		panic(err)
	}

	// Start send audio goroutine
	sendAudioDone := make(chan error, 1)

	go func() {
		sendAudioDone <- func() error {
			// Send audio data
			buffer := make([]byte, 3840)

			for {
				n, err := file.Read(buffer)
				if err != nil && err != io.EOF {
					return err
				}
				if n == 0 {
					break
				}

				if err := conn.WriteMessage(websocket.BinaryMessage, buffer[:n]); err != nil {
					return err
				}

				// Sleep for 120 ms
				time.Sleep(120 * time.Millisecond)
			}

			// Send end of file
			if err := conn.WriteMessage(websocket.TextMessage, []byte{}); err != nil {
				return err
			}

			return nil
		}()
	}()

	fmt.Println("Transcription started.")

	// Receive text
	finalText := ""

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
				text := err.(*websocket.CloseError).Text
				if text != "" {
					fmt.Printf("Close error: %s\n", text)
				}
				break
			}
			panic(err)
		}

		var res Response
		if err := json.Unmarshal(message, &res); err != nil {
			panic(err)
		}
		if res.ErrorCode != 0 {
			fmt.Printf("\nError: %d %s\n", res.ErrorCode, res.ErrorMessage)
			break
		}

		nonFinalText := ""

		for _, token := range res.Tokens {
			if token.Text != "" {
				if token.IsFinal {
					finalText += token.Text
				} else {
					nonFinalText += token.Text
				}
			}
		}

		fmt.Println(
			"\033[2J\033[H" + // clear the screen, move to top-left corner
				finalText + // write final text
				"\033[34m" + // change text color to blue
				nonFinalText + // write non-final text
				"\033[39m", // reset text color
		)

		if res.Finished {
			fmt.Println("\nTranscription done.")
		}
	}

	if err := <-sendAudioDone; err != nil {
		panic(err)
	}
}
